scheduler/distributed_task_test.go

```diff
@@ -0,0 +1,156 @@
+	"time"
+
+	"github.com/intelsdi-x/snap/control"
+	"github.com/intelsdi-x/snap/core"
+	"github.com/intelsdi-x/snap/pkg/schedule"
+	"github.com/intelsdi-x/snap/scheduler/wmap"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+var (
+	PluginPath = path.Join(SnapPath, "plugin")
+)
+
+func TestDistributedWorkflow(t *testing.T) {
+	Convey("Create a scheduler with 2 controls and load plugins", t, func() {
+		l, _ := net.Listen("tcp", ":0")
+		l.Close()
+		cfg := control.GetDefaultConfig()
+		cfg.ListenPort = l.Addr().(*net.TCPAddr).Port
+		c1 := control.New(cfg)
+		c1.Start()
+		m, _ := net.Listen("tcp", ":0")
+		m.Close()
+		cfg.ListenPort = m.Addr().(*net.TCPAddr).Port
+		port1 := cfg.ListenPort
+		c2 := control.New(cfg)
+		n, _ := net.Listen("tcp", ":0")
+		n.Close()
+		schcfg := GetDefaultConfig()
+		schcfg.ListenPort = n.Addr().(*net.TCPAddr).Port
+		sch := New(schcfg)
+		c2.Start()
+		sch.SetMetricManager(c1)
+		err := sch.Start()
+		So(err, ShouldBeNil)
+		// Load appropriate plugins into each control.
+		mock2Path := path.Join(PluginPath, "snap-collector-mock2")
+		passthruPath := path.Join(PluginPath, "snap-processor-passthru")
+		filePath := path.Join(PluginPath, "snap-publisher-file")
+
+		// mock2 and file onto c1
+
+		rp, err := core.NewRequestedPlugin(mock2Path)
+		So(err, ShouldBeNil)
+		_, err = c1.Load(rp)
+		So(err, ShouldBeNil)
+		rp, err = core.NewRequestedPlugin(filePath)
+		So(err, ShouldBeNil)
+		_, err = c1.Load(rp)
+		So(err, ShouldBeNil)
+		// passthru on c2
+		rp, err = core.NewRequestedPlugin(passthruPath)
+		So(err, ShouldBeNil)
+		passthru, err := c2.Load(rp)
+		So(err, ShouldBeNil)
+
+		Convey("Test task with one local and one remote node", func() {
+			//Create a task
+			//Create a workflowmap
+			wf := dsWFMap(port1)
+			t, errs := sch.CreateTask(schedule.NewSimpleSchedule(time.Second), wf, true)
+			So(len(errs.Errors()), ShouldEqual, 0)
+			So(t, ShouldNotBeNil)
+		})
+
+		Convey("Test task with invalid remote port", func() {
+			wf := dsWFMap(0)
+			t, errs := sch.CreateTask(schedule.NewSimpleSchedule(time.Second), wf, true)
+			So(len(errs.Errors()), ShouldEqual, 1)
+			So(t, ShouldBeNil)
+		})
+
+		Convey("Test task without remote plugin", func() {
+			_, err := c2.Unload(passthru)
+			So(err, ShouldBeNil)
+			wf := dsWFMap(port1)
+			t, errs := sch.CreateTask(schedule.NewSimpleSchedule(time.Second), wf, true)
+			So(len(errs.Errors()), ShouldEqual, 1)
+			So(t, ShouldBeNil)
+		})
+
+		Convey("Test task failing when control is stopped while task is running", func() {
+			wf := dsWFMap(port1)
+			interval := time.Millisecond * 100
+			t, errs := sch.CreateTask(schedule.NewSimpleSchedule(interval), wf, true)
+			So(len(errs.Errors()), ShouldEqual, 0)
+			So(t, ShouldNotBeNil)
+			c2.Stop()
+			// Give task time to fail
+			time.Sleep(time.Second)
+			tasks := sch.GetTasks()
+			var task core.Task
+			for _, v := range tasks {
+				task = v
+			}
+			So(task.State(), ShouldEqual, core.TaskDisabled)
+		})
```

Comment:
Consider adding tests for subscriptions (local and remote) in cases where the task is successful and where deps fail and we need to rollback.